### PR TITLE
Remove HashableObjectStorage

### PR DESCRIPTION
**Pull request by Benestar**
HashableObjectStorage was a package-private base class which was
only used in ReferenceList. This patch just removes this inheritance,
adds all methods to ReferenceList and removes HashableObjectStorage.
More refactoring in ReferenceList will be done in a follow-up patch.

See #212

### CODE REVIEW

File: src/ReferenceList.php [hunk 1]
```diff
@@ -23,7 +27,7 @@
  * @author H. Snater < mediawiki@snater.com >
  * @author Thiemo Mättig
  */
-class ReferenceList extends HashableObjectStorage {
+class ReferenceList extends SplObjectStorage implements Comparable {
```
Comment (thiemowmde) [inline src/ReferenceList.php]: I'm not sure if this is an improvement. I like it more if that bogus SPL stuff is in an other class.

File: src/ReferenceList.php [hunk 2]
```diff
@@ -241,4 +245,59 @@ public function unserialize( $data ) {
 		$this->__construct( unserialize( $data ) );
 	}
 
+	/**
+	 * Removes duplicates bases on hash value.
+	 *
+	 * @since 0.2
+	 */
+	public function removeDuplicates() {
```
Comment (thiemowmde) [inline src/ReferenceList.php]: Is this even need in any code base?

File: src/ReferenceList.php [hunk 3]
```diff
@@ -241,4 +245,59 @@ public function unserialize( $data ) {
...
+	}
+
+	/**
+	 * The hash is purely valuer based. Order of the elements in the array is not held into account.
+	 *
+	 * @since 0.3
+	 *
+	 * @return string
+	 */
+	public function getValueHash() {
```
Comment (thiemowmde) [inline src/ReferenceList.php]: Does this need to be public? Where is this used outside of this class?

File: src/ReferenceList.php (hunk 1 above)
Comment (Benestar) [inline src/ReferenceList.php]: I decided to do it this way round (first remove `HashableObjectStorage`, then refactoring the class). If you prefer it doing the other way round (first removing the SPL stuff) and have some arguments I can also do it that way.

File: src/ReferenceList.php (hunk 2 above)
Comment (Benestar) [inline src/ReferenceList.php]: No, we should remove this method (and many others). However, I'm not sure if our next release should be a breaking release again. :S

File: src/ReferenceList.php (hunk 3 above)
Comment (Benestar) [inline src/ReferenceList.php]: This is called once in `Statement`. However, it should call the `MapValueHasher` there and this method should be removed as well.

**Comment by thiemowmde** [comment]
I understand that this patch is meant to be as simple as possible, to just move stuff without changing the public interface of the ReferenceList class. But I'm not sure if this is worth the trouble and our combined developers time. This is not in a sprint at the moment.

I don't want to block this patch, but I will not +1 it at the moment.

**Comment by Benestar** [comment]
@thiemowmde You don't have to review this patch in the current sprint if you don't have time. I'm just preparing some stuff, maybe for the next sprint. These classes are imo very unlikely to be modified next time so there won't be any merge conflicts.

**Comment by thiemowmde** [comment]
I know, everything is fine. :-) Go on.

File: src/ReferenceList.php (hunk 2 above)
Comment (thiemowmde) [inline src/ReferenceList.php]: Good point.

File: src/ReferenceList.php (hunk 1 above)
Comment (thiemowmde) [inline src/ReferenceList.php]: Go on. You can add an other commit (in an other pull request) on top of this.

**Comment by JeroenDeDauw** [comment]
@Benestar How do you figure this thing is package private?

**Comment by Benestar** [comment]
I assumed they were but apperently I just hoped that :P So this is going 
to be a breaking change anyways, right? I'm not a fan of doing another 
major release after we just did one though. :S What are your thoughts?

Am 11.06.2015 um 16:28 schrieb Jeroen De Dauw:

**Comment by JeroenDeDauw** [comment]
Just keep this open for the next breaking release. Doing two close together is not an issue, but at this point we simple do not have anything that warrants one.

**Comment by Benestar** [comment]
Sounds good, so I will just continue working on this branch and when we are ready for the next major release we merge them. This issue isn't urgent, just nice to have. ;-)

**Comment by Benestar** [comment]
Anything still blocking this? @JeroenDeDauw @thiemowmde @addshore

**Comment by JeroenDeDauw** [comment]
Meh, looks like we forgot to include this in 4.0? >_> Well, same argumentation still applies, let's wait till we actually have sufficient stuff to warrant a new breaking release. Though let's try to not forget it this time...

**Comment by JeroenDeDauw** [comment]
PHPMD agrees we should remove this class

> The class ReferenceList has 11 public methods. Consider refactoring ReferenceList to keep number of public methods under 10.

:)

**Comment by JeroenDeDauw** [comment]
Needs rebase

File: src/ReferenceList.php [hunk 4]
```diff
@@ -55,31 +62,23 @@ public function __construct( $references = array() ) {
 	 * @throws InvalidArgumentException
 	 */
 	public function addReference( Reference $reference, $index = null ) {
-		if ( !is_int( $index ) && $index !== null ) {
-			throw new InvalidArgumentException( '$index must be an integer or null' );
+		if ( $index !== null && ( !is_int( $index ) || $index < 0 ) ) {
+			throw new InvalidArgumentException( '$index must be a non-negative integer or null' );
```
Comment (thiemowmde) [inline src/ReferenceList.php]: The non-negative check is a breaking change.

File: src/ReferenceList.php [hunk 5]
```diff
@@ -46,6 +52,7 @@ public function __construct( $references = array() ) {
 
 	/**
 	 * Adds the provided reference to the list.
+	 * Empty references are ignored.
```
Comment (thiemowmde) [inline src/ReferenceList.php]: This is a breaking change.

File: src/ReferenceList.php [hunk 6]
```diff
@@ -220,25 +177,94 @@ public function getReference( $referenceHash ) {
...
-	 * @see Serializable::serialize
+	 * Returns the index of a reference or false if the reference could not be found.
+	 *
+	 * @since 0.5
+	 *
+	 * @param Reference $reference
+	 *
+	 * @return int|boolean
+	 */
+	public function indexOf( Reference $reference ) {
```
Comment (thiemowmde) [inline src/ReferenceList.php]: This should be moved up so the diff is more minimal.

File: src/ReferenceList.php [hunk 7]
```diff
@@ -220,25 +177,94 @@ public function getReference( $referenceHash ) {
...
+	public function getHash() {
+		$hashes = array();
+
+		foreach ( $this->references as $reference ) {
+			$hashes[] = $reference->getHash();
+		}
+
+		sort( $hashes );
+
+		return implode( '|', $hashes );
```
Comment (thiemowmde) [inline src/ReferenceList.php]: This is a breaking change. The old algorithm used MapValueHasher, which used sha1. This one does not, possibly creating strings that are thousands of characters long.

File: src/Statement/Statement.php [hunk 8]
```diff
@@ -231,7 +230,7 @@ public function getHash() {
 			array(
 				sha1( $this->mainSnak->getHash() . $this->qualifiers->getHash() ),
 				$this->rank,
-				$this->references->getValueHash(),
+				$this->references->getHash(),
```
Comment (thiemowmde) [inline src/Statement/Statement.php]: This must also be announced as a breaking change.

**Comment by thiemowmde** [comment]
I just tried to rebase this patch, but am afraid of the conflicts and possible issues I see. Why are there two commits in here, that do have their own pull requests (#500 and #501) that are already merged? Does this get lost when I rebase and squish this?

Why are there so many unrelated breaking changes in this patch? I tried to mark them all with comments. I think this should be split and discussed separately.

I'm mostly pro this change and most of the breaking changes it includes. I would love to see this moving forward, but I do not know exactly how we could do that. At the moment it seems easier to resubmit a series of much smaller changes instead of continuing this one.

**Comment by Benestar** [comment]
Hah, I tried to resolve this branching and merging weirdness but apparently only messed around even more -.- Perhaps it's the easiest way to create clean new branches with the changes applied?

**Pull request closed without merging**